### PR TITLE
Fix sorting issues on `x509.certificate_managed` triggering unnecessary changes

### LINKED ISSUE #56556: x509.certificate_managed triggers a change when switching form python 2.7 to 3.6
### Description of Issue
While testing salt-3000.1 on py36 vs on py27 I noticed all my certificates are regenerated, including my CA! (this is absolutely not what I want, as the CA is also use for things outside of salt)

I restored backups of the 10 effected minions and double checked I was on py27 salt-3000.1 which I was, another global state.apply showed no changes, phew!

**Switching to from python 2.7 to 3.6 for salt-3000.1 results in all x509.certificate_managed states to regenerate certificates**

### Debugging done
Debugging is hard as I could not leave everything in this broken state for long.

Using pudb i stepped through the x509 state and module:
- [state] x509.certificate_managed's `current_comp == new_comp` comparison was false
- [state] x509.certificate_managed's `current_comp == new_comp` had different `Subject Hash` fields
- [module] x509.read_certificate sets this hash using: `_dec2hex(cert.get_subject().as_hash())`

Stripping this down to the value of just `cert.get_subject()`:
 - py27:
```
/C=BE/CN=Example Root CA/L=Kapellen/ST=Example/O=Example/emailAddress=certadm@example.org
```
- py36:
```
/C=BE/ST=Antwerp/L=Kapellen/O=Example/CN=Example Root CA/emailAddress=certadm@example.org
```

The ordering seems to be different between both version of python (but consistant within the version itself)

### Setup
Heavily stripped down state due to sensitive info
```
certificate.authority::directory:
  file.directory:
    - name: {{ certcfg['config']['authority_dir'] }}
    - makedirs: true
    - user: root
    - group: nacl
    - dir_mode: 2770

certificate.authority::private-key:
  x509.private_key_managed:
    - name: {{ ca_key_path }}
    - bits: 8192
    - backup: True
    - require:
      - file: certificate.authority::directory

certificate.authority::certificate:
  x509.certificate_managed:
    - name: {{ ca_crt_path }}
    - signing_private_key: {{ ca_key_path }}
    - CN: Example Root CA
    - O: Example
    - C: BE
    - ST: Antwerp
    - L: Kapellen
    - Email: certadm@example.org
    - basicConstraints: "critical CA:true"
    - keyUsage: "critical cRLSign, keyCertSign"
    - subjectKeyIdentifier: hash
    - authorityKeyIdentifier: keyid,issuer:always
    - days_valid: 3650
    - days_remaining: 0
    - backup: True
    - require:
      - x509: certificate.authority::private-key
```

### Steps to Reproduce Issue
- setup a py27 based salt-3000.1 master/minion
- manage a private key + ca cert
- manage a cert signed by the cs
- apply the state
- upgrade to py36 based salt-3000.1 master/minion
- apply the state
- notice both the ca and signed certificate change

### Versions Report
```
[root@cronos /data]# salt-call --versions-report
Salt Version:
           Salt: 3000.1

Dependency Versions:
           cffi: 1.11.5
       cherrypy: unknown
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: 0.30.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 3.6.6
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 3.6.6 (default, Oct 23 2018, 17:13:47)
   python-gnupg: Not Installed
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist:
         locale: UTF-8
        machine: i86pc
        release: 5.11
         system: SunOS
        version: Not Installed
```

```
[root@cronos /data]# salt-call --versions-report
Salt Version:
           Salt: 3000.1

Dependency Versions:
           cffi: 1.11.5
       cherrypy: 17.4.0
       dateutil: 2.8.1
      docker-py: Not Installed
          gitdb: 2.0.6
      gitpython: 2.1.14
         Jinja2: 2.10
        libgit2: Not Installed
       M2Crypto: 0.30.1
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.5.6
   mysql-python: Not Installed
      pycparser: 2.19
       pycrypto: 3.7.2
   pycryptodome: Not Installed
         pygit2: Not Installed
         Python: 2.7.15 (default, Sep 10 2019, 16:35:50)
   python-gnupg: 0.4.5
         PyYAML: 3.13
          PyZMQ: 17.1.2
          smmap: 0.9.0
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.2.5

System Versions:
           dist:
         locale: UTF-8
        machine: i86pc
        release: 5.11
         system: SunOS
        version: Not Installed
```

### PR DESCRIPTION
**Pull request by s0undt3ch**
### What does this PR do?
See title

### What issues does this PR fix or reference?
Fixes #56556 
Closes #56788

### Previous Behavior
The certificate would be regenerated because of sorting differences between Py2 and Py3

### New Behavior
The certificate is no longer, unnecessarily regenerated

### Merge requirements satisfied?
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

**Review by krionbsd — APPROVED** [review]

**Comment by sjorge** [comment]
Just tested this change and it fixes my issue as reported in #56556

**Comment by s0undt3ch** [comment]
/CC @OrangeDog @hatifnatt @glynnforrest @alxwr
You were all involved on either the issue being fixed on this PR or a linked issue.
Does this address your concerns? See any problematic code in this solution?

**Comment by OrangeDog** [comment]
It doesn't seem to have worked. Wants to change it every run.
```
local:
----------
          ID: /etc/ssl/private/example.key
    Function: x509.private_key_managed
      Result: None
     Comment: The file /etc/ssl/private/example.key is set to be changed
              Note: No changes made, actual changes may
              be different due to other states.
     Started: 15:26:23.064729
    Duration: 54.638 ms
     Changes:
              ----------
              new:
                  New private key generated
----------
          ID: /etc/ssl/certs/example.crt
    Function: x509.certificate_managed
      Result: None
     Comment: Certificate /etc/ssl/certs/example.crt will be created
     Started: 15:26:23.121432
    Duration: 366.75 ms
     Changes:
              ----------
              Status:
                  ----------
                  New:
                      Certificate will be valid and up to date
                  Old:
                      Certificate properties are different: Public Key
----------
```
```sls
{{ pki.key_path }}:
  x509.private_key_managed:
    - bits: 2048
    - group: {{ pki.group }}
    - mode: 640
    - new: true
    - require:
      - sls: pki.base
  {% if salt['file.file_exists'](pki.key_path) %}
    - prereq:
      - x509: {{ pki.cert_path }}
  {% endif %}

{{ pki.cert_path }}:
  x509.certificate_managed:
    - ca_server: salt
    - signing_policy: internal
    - public_key: {{ pki.key_path }}
    - days_remaining: 7
    - CN: {{ grains['fqdn'] }}
    - subjectAltName: "DNS:{{ grains['fqdn'] }}, DNS:{{ grains['host'] }}"
    - require:
      - pkg: openssl
```

**Comment by hatifnatt** [comment]
@s0undt3ch sorry can't test this for now because it will require setting up a lot of things.

**Comment by s0undt3ch** [comment]
@hatifnatt sure, no problem.
@OrangeDog will take your example for a spin.

**Comment by s0undt3ch** [comment]
Ultimately, but for 3002, we should, as @OrangeDog mentioned before, inspect/parse the existing certificate and compare the fields instead of generating a new temporary certificate just to compare.
Anyone wants to grab that task?
My certificate knowledge can expand but is scarse and might take me more time than any of you.

**Comment by glynnforrest** [comment]
> Anyone wants to grab that task?

I was hoping to give it a go after https://github.com/saltstack/salt/pull/56788 was merged, I've got a rough idea how it would work after #52935.

I haven't tested this yet, but appreciate the cleanup of six usage. Even if this doesn't fix everything, it seems like a step in the right direction :+1:

File: salt/states/x509.py [hunk 1]
```diff
@@ -170,27 +170,24 @@
...
 
-# Import 3rd-party libs
-from salt.ext import six
-
 try:
     from M2Crypto.RSA import RSAError
 except ImportError:
     pass
 
+log = logging.getLogger(__name__)
```
Comment (glynnforrest) [inline salt/states/x509.py:189]: This appears to be unused? I guess you wrote debug logs while working on it.

Comment (s0undt3ch) [reply salt/states/x509.py:189]: Yeah, I did. It's "inoffensive", we can leave it, unless you object :)

**Comment by s0undt3ch** [comment]
@glynnforrest, FYI, #56788 was cherry-picked into this PR.

**Comment by OrangeDog** [comment]
There don't seem to have been any changes since I confirmed this didn't work.
Did nobody reproduce it or something?
Why was this merged?

**Comment by sjorge** [comment]
> There don't seem to have been any changes since I confirmed this didn't work.
> Did nobody reproduce it or something?
> Why was this merged?

Last I tried with this change it will also s till broken for me, but 3001 was so broken on my paltform of choice I haven't had time to retest.

**Comment by s0undt3ch** [comment]
This *did* partially addressed the problem, at least with the example given at the time and I did not have time to test(at fix) with your "dataset". The *right* fix is by not doing this kind of checks or generating a temp certificate to compare. The *right* fix would not(and will not) land in 3001.1.

**Comment by OrangeDog** [comment]
The linked issue(s) need to be reopened then because they're not fixed.

**Comment by s0undt3ch** [comment]
Yep, I'm reopening it.

**Comment by alxwr** [comment]
@s0undt3ch Thanks for reopening!
I was able to reproduce the bug on 3001.1 and found a way to solve (or work around) this issue. At it's core it seems to be just a type mismatch:

```python3
>>> b'asdf' == 'asdf'
False
>>> type(b'asdf')
<class 'bytes'>
>>> type('asdf')
<class 'str'>
>>> b'asdf' == str.encode('asdf')
True
```

This patch worked for me on FreeBSD 11.3-RELEASE-p12 using Salt 3001.1.

```diff
--- a/salt/states/x509.py   2020-08-23 23:01:57.401782000 +0200
+++ b/salt/states/x509.py   2020-08-23 23:20:37.222170000 +0200
@@ -465,6 +465,9 @@
                         pass
                     if v_ext != cert_info_ext:
                         diff.append(k)
+                elif type(v) == str and type(cert_info[k]) == bytes:
+                    if str.encode(v) != cert_info[k]:
+                        diff.append(k)
                 else:
                     diff.append(k)
         except KeyError:f

```
I'm currently working on a different project and don't have much time. Sry. Furthermore I'm not familiar with Salt core development. So my patch may be full of bad ideas. :-)
Feel free to use my experiment for a proper PR. (I'll open a PR if you want to, but please don't expect regular updates.) I just wanted to share my findings in case they are useful for somebody.

----

Versions report:

```
Salt Version:
           Salt: 3001.1
 
Dependency Versions:
           cffi: 1.14.2
       cherrypy: Not Installed
       dateutil: Not Installed
      docker-py: Not Installed
          gitdb: Not Installed
      gitpython: Not Installed
         Jinja2: 2.10.1
        libgit2: Not Installed
       M2Crypto: 0.36.0
           Mako: Not Installed
   msgpack-pure: Not Installed
 msgpack-python: 0.6.2
   mysql-python: Not Installed
      pycparser: 2.20
       pycrypto: Not Installed
   pycryptodome: 3.9.8
         pygit2: Not Installed
         Python: 3.7.8 (default, Aug 17 2020, 08:38:11)
   python-gnupg: Not Installed
         PyYAML: 5.3.1
          PyZMQ: 19.0.2
          smmap: Not Installed
        timelib: Not Installed
        Tornado: 4.5.3
            ZMQ: 4.3.1
 
System Versions:
           dist: freebsd 11.3 
         locale: UTF-8
        machine: amd64
        release: 11.3-RELEASE-p12
         system: FreeBSD
        version: FreeBSD 11.3
```

**Comment by alxwr** [comment]
I turned my last comment into a PR: #58296

**Pull request merged**